### PR TITLE
alert 5.0.0 upgrade environs support, operator ui 2019.8.1 Black Duck support, scale down wait for pods to be deleted

### PR DESCRIPTION
**Pull request by msenmurugan**
* operator ui support for 2019.8.1
* synopsysctl upgrade alert support for environs migrations
* wait for pod count to replicas count during scale down

### CODE REVIEW

**Comment by operator-r2d2** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/blackducksoftware/synopsys-operator/pull/895#" title="Author self-approved">msenmurugan</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=blackducksoftware%2Fsynopsys-operator).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/blackducksoftware/synopsys-operator/blob/release-2019.8.x/OWNERS)~~ [msenmurugan]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: pkg/synopsysctl/cmd_update.go [hunk 1]
```diff
@@ -386,6 +386,41 @@ func updateAlert(alt *alertapi.Alert, flagset *pflag.FlagSet) (*alertapi.Alert,
...
+	if err != nil {
+		return nil, err
+	}
+
+	// if greater than or equal to 5.0.0, then copy PUBLIC_HUB_WEBSERVER_HOST to ALERT_HOSTNAME and PUBLIC_HUB_WEBSERVER_PORT to ALERT_SERVER_PORT
+	// and delete PUBLIC_HUB_WEBSERVER_HOST and PUBLIC_HUB_WEBSERVER_PORT from the environs. In future, we need to request the customer to use the new params
+	if isGreaterThanOrEqualTo {
+		isChanged := false
+		maps := util.StringArrayToMapSplitBySeparator(newSpec.Environs, ":")
+		_, ok := maps["PUBLIC_HUB_WEBSERVER_HOST"]
```
Comment (rrati) [inline pkg/synopsysctl/cmd_update.go]: I think we can simply this logic thusly:

if _, ok := maps["PUBLIC_HUB_WEBSERVER_HOST"]; ok {
  if _, ok1 := maps["ALERT_HOSTNAME"]; !ok1 {
    maps["ALERT_HOSTNAME"] = maps["PUBLIC_HUB_WEBSERVER_HOST"]
  }
  delete(maps, "PUBLIC_HUB_WEBSERVER_HOST")
  isChanged = true
}

Same logic for PORT

Then lastly re-do the conversion:
if isChanged {
  newSpec.Environs = util.MapToStringArrayJoinBySeparator(maps, ":")
}

File: pkg/util/common.go [hunk 2]
```diff
@@ -1324,6 +1349,17 @@ func PatchReplicationControllerForReplicas(clientset *kubernetes.Clientset, old
 	if err != nil {
 		return nil, err
 	}
+
+	for {
```
Comment (rrati) [inline pkg/util/common.go]: Should have a reasonable timeout rather than loop forever

**Comment by operator-r2d2** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/blackducksoftware/synopsys-operator/pull/895#" title="Author self-approved">msenmurugan</a>*, *<a href="https://github.com/blackducksoftware/synopsys-operator/pull/895#pullrequestreview-293128467" title="Approved">rrati</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=blackducksoftware%2Fsynopsys-operator).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/blackducksoftware/synopsys-operator/blob/release-2019.8.x/OWNERS)~~ [msenmurugan,rrati]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: pkg/synopsysctl/cmd_update.go (hunk 1 above)
Comment (msenmurugan) [reply pkg/synopsysctl/cmd_update.go]: done

File: pkg/util/common.go (hunk 2 above)
Comment (msenmurugan) [reply pkg/util/common.go]: done

**Comment by operator-r2d2** [comment]
@msenmurugan: once the present PR merges, I will cherry-pick it on top of release-2019.10.x in a new PR and assign it to you.

<details>

In response to [this](https://github.com/blackducksoftware/synopsys-operator/pull/895#issuecomment-535091861):

>/cherrypick release-2019.10.x


Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

**Comment by operator-r2d2** [comment]
@msenmurugan: #895 failed to apply on top of branch "release-2019.10.x":
```Applying: alert 5.0.0 migrate support
error: Failed to merge in the changes.
Using index info to reconstruct a base tree...
M	cmd/operator-ui/go.mod
M	go.mod
M	go.sum
M	pkg/synopsysctl/cmd_update.go
M	pkg/util/common.go
Falling back to patching base and 3-way merge...
Auto-merging pkg/util/common.go
Auto-merging pkg/synopsysctl/cmd_update.go
Auto-merging go.sum
CONFLICT (content): Merge conflict in go.sum
Auto-merging go.mod
CONFLICT (content): Merge conflict in go.mod
Auto-merging cmd/operator-ui/go.mod
Patch failed at 0001 alert 5.0.0 migrate support

```

<details>

In response to [this](https://github.com/blackducksoftware/synopsys-operator/pull/895#issuecomment-535091861):

>/cherrypick release-2019.10.x


Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

**Pull request merged**